src/libraries/System.Private.Uri/src/System/UncNameHelper.cs

```diff
@@ -49,34 +49,35 @@ internal static string ParseCanonicalName(string str, int start, int end, ref bo
         //
         // Assumption is the caller will check on the resulting name length
         // Remarks:  MUST NOT be used unless all input indexes are verified and trusted.
-        internal static unsafe bool IsValid(char* name, ushort start, ref int returnedEnd, bool notImplicitFile)
+        internal static unsafe bool IsValid(char* name, int start, ref int returnedEnd, bool notImplicitFile)
         {
-            ushort end = (ushort)returnedEnd;
+            int end = returnedEnd;
 
             if (start == end)
                 return false;
             //
             // First segment could consist of only '_' or '-' but it cannot be all digits or empty
             //
             bool validShortName = false;
-            ushort i = start;
+            int i = start;
             for (; i < end; ++i)
             {
-                if (name[i] == '/' || name[i] == '\\' || (notImplicitFile && (name[i] == ':' || name[i] == '?' || name[i] == '#')))
+                char c = name[i];
+                if (c == '/' || c == '\\' || (notImplicitFile && (c == ':' || c == '?' || c == '#')))
                 {
                     end = i;
                     break;
                 }
-                else if (name[i] == '.')
+                else if (c == '.')
                 {
                     ++i;
                     break;
                 }
-                if (char.IsLetter(name[i]) || name[i] == '-' || name[i] == '_')
+                if (char.IsLetter(c) || c == '-' || c == '_')
                 {
                     validShortName = true;
                 }
-                else if (name[i] < '0' || name[i] > '9')
+                else if (c < '0' || c > '9')
```

Comment:
`c - '0' > '9' - '0'` ?

Reply:
Saw this as well, I am planning on changing similar checks as a part of a different PR in the uri-cleanup series.

Reply:
I think we need to make `char.IsInRange` public for that. The hack I suggested is widely used in the BCL but looks ugly. I had a draft PR to JIT to handle this pattern there https://github.com/dotnet/coreclr/pull/27480

Reply:
I was thinking of isolating common character checks into a Uri internal `CharHelper` class like
```c#
internal static class CharHelper
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsAsciiDigit(char c) =>
        (uint)(c - '0') <= ('9' - '0');

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsAsciiLowercaseLetter(char c) =>
        (uint)(c - 'a') <= ('z' - 'a');

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsAsciiUppercaseLetter(char c) =>
        (uint)(c - 'A') <= ('Z' - 'A');

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsAsciiLetter(char c) =>
        (uint)((c - 'A') & ~0x20) <= ('Z' - 'A');

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsAsciiLetterOrDigit(char c) =>
        ((uint)((c - 'A') & ~0x20) <= ('Z' - 'A')) ||
        ((uint)(c - '0') <= ('9' - '0'));

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsHexDigit(char c) =>
        ((uint)(c - '0') <= ('9' - '0')) ||
        ((uint)((c - 'A') & ~0x20) <= ('F' - 'A'));

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool IsInInclusiveRange(char c, char min, char max) =>
        (uint)(c - min) <= (uint)(max - min);
}
```